pkg/app/ops/insightcollector/mode.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2021 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package insightcollector
+
+import (
+	"bytes"
+)
+
+type CollectorMode uint
```

Comment:
@sanposhiho Why do we need this type instead of using the config value directly?

Reply:
Hmm, I've thought about that too.

IMO, I don't want config and insightcollector to be closely related.
I'd like to implement specific behavior in the insightcollector itself, so that the insightcollector doesn't directly change its behavior depending on external values. (In other words, the logic of insightcollector should be completed in insightcollector, and only the interface provided by insightcollector should be used from outside.

In this implementation, the feature-flag of insight is managed in config. But, for example,  we may change it in the future to manage the feature-flag of insight from another place. Whenever we make such a change, we will need to make internal changes to the insightcollector, if we directly use config value to decide the behavior of insightcollector.  If we provide CollectorMode externally, and implement insightcollector to change its behavior only depending on CollectorMode, we only need to change the usage of CollectorMode for the change.

How do you think?

Reply:
@sanposhiho I got your points. Really nice thoughts. 👍 
I just wanted to keep this as simple as possible.
So let's keep this Mode definition as your suggestion, but let me leave other comments to simplify around this part.